### ISSUE TITLE
docker task doesn't respect configuration scopes

### ISSUE THREAD

**Issue by RomanIakovlev** [marcuslonnberg/sbt-docker#44]
... or so it seems.

Background: I'm writing a SBT plugin for developers to run project's SBT session inside Docker container. This is needed because our application needs a specific Linux environment, and some developers use OSX. Something like Vagrant. I also want to package the application to a Docker image for deployment. Docker image for development and for publishing are different. I want to use `sbt-docker` for both of these tasks.

But when I want to scope `dockerfile` setting a config other than Global, I run into this issue. If I define `dockerfile in docker` only inside some specific configuration, I can't build the image. I get the following error:

 ```
(*:docker::dockerfile) A Dockerfile is not defined. Please define one with `dockerfile in docker`
```
even though I have the following settings in my plugin:
```
object autoImport {
  lazy val prepareDevContainer = taskKey[String]("Prepares a Docker image to run the SBT session in")
  lazy val devContainerConfig = config("devcontainer") extend(Compile)
}

private lazy val devContainerSettings = Seq(
    dockerfile in docker := createDevDockerfile(), // details elided
    prepareDevContainer := {
      val port = play.PlayImport.PlayKeys.playDefaultPort.value
      s"docker run -it -p $port:$port -v ${ivyPaths.value.ivyHome.get}:/tmp/.ivy2 " +
        s"-v ${System.getProperty("user.home")}/.aws/config:/root/.aws/config " +
        s"-v ${baseDirectory.value.getParent}:/src ${DockerKeys.docker.value.id} " +
        "/bin/sh -c \"cd application; sbt -Dsbt.ivy.home=/tmp/.ivy2 -Divy.home=/tmp/.ivy2\""
    },
    imageNames in docker := Seq(ImageName("sbt-centos7")))

override def projectSettings: Seq[Def.Setting[_]] = 
  inConfig(devContainerConfig)(devContainerSettings)
```

If I change the `projectSettings` definitions to the following:
```
override def projectSettings: Seq[Def.Setting[_]] = 
  inConfig(devContainerConfig)(devContainerSettings) ++ Seq(
    dockerfile in docker := createProdDockerfile())
```
then I can build the image, but I can't use different `dockerfile`s for dev and production, and `createProdDockerfile()` value is always used.

In other words, in sbt console `show docker::dockerfile` and `show devcontainer:docker::dockerfile` correctly give different output, but executing `devcontainer:docker` incorrectly produce the same image as just `docker`.

**Comment by romanwozniak** [comment]
Oh yes! I get stuck with this issue. I'm new to SBT so I thought that I'm doing something wrong...

I'm trying to build two different docker images (that differ one from other by CMD line) depending on configuration scope. But it looks to be impossible.

**Comment by marcuslonnberg** [comment]
I haven't tried this myself before but I added a test for it now which seems to work.
Take a look here: 3ff4774

@RomanIakovlev it might work for you if you add this to `projectSettings`:
`inConfig(devContainerSettings)(sbtdocker.DockerSettings.baseDockerSettings)`

Please tell me if you get it to work or not.

**Comment by RomanIakovlev** [comment]
@marcuslonnberg sorry it took so long for me to come back to this issue. I still haven't solved it. I've created a minimal example to illustrate this issue. I'd appreciate if you have a look at it and tell what's wrong there. https://github.com/RomanIakovlev/sbt-docker-test

**Comment by marcuslonnberg** [comment]
Great with the example project.

Is there a reason why you want to scope the settings by task and not by config?
I don't know if you can scope a setting on a task on a task.
If you would change `dockerTask` to a config and use `inConfig` instead of `inTask` it should probably work.

**Comment by RomanIakovlev** [comment]
As far as I know, from SBT perspective there should be no difference between scoping on tasks, configs or projects. The mechanism is the same. For my real use case, I'd prefer to keep it simple and don't introduce configurations when a task would suffice. Initial example mentioned configuration scoping just because I happened to experiment with configuration at that moment. I'll try a minimal example with configuration and let you know.

**Comment by RomanIakovlev** [comment]
Yep, I confirm it works when I change to configuration as you suggest. You can find the updated example in the branch https://github.com/RomanIakovlev/sbt-docker-test/tree/config-scoping. However, I still feel task scoping should be supported. Imagine having different images for run and test. 

But maybe I misunderstand how SBT scoping works. I know for sure you can scope settings on tasks, but in this case one task (`docker`) execution is triggered by another task (`dockerTask` in my minimal example), and maybe scoped setting value (`dockerfile in dockerTask`) is not transferred from the original task.

**Comment by marcuslonnberg** [comment]
I think you have 2 options at the moment.
You can use configurations, either the built-in like `Test` or define you own.

Or you can define new tasks, replacing `docker`, `dockerPush` and `dockerBuildAndPush` with some other names:
```scala
docker := {
  val log = Keys.streams.value.log
  val dockerPath = (DockerKeys.dockerPath in docker).value
  val buildOptions = (DockerKeys.buildOptions in docker).value
  val stageDir = (target in docker).value
  val dockerfile = (DockerKeys.dockerfile in docker).value
  val imageNames = (DockerKeys.imageNames in docker).value
  DockerBuild(dockerfile, DefaultDockerfileProcessor, imageNames, buildOptions, stageDir, dockerPath, log)
},
dockerPush := {
  val log = Keys.streams.value.log
  val dockerPath = (DockerKeys.dockerPath in docker).value
  val imageNames = (DockerKeys.imageNames in docker).value

  DockerPush(dockerPath, imageNames, log)
},
dockerBuildAndPush <<= (docker, dockerPush) { (build, push) =>
  build.flatMap { id =>
    push.map(_ => id)
  }
}
```

See `sbtdocker.DockerSettings`.

Does that solve your problem?

**Comment by RomanIakovlev** [comment]
I've settled for configuration at the moment. Thanks for the clarifications! I close the issue.
